### ISSUE TITLE
Advice Needed - Mouse History Rewind Feature?

### ISSUE THREAD

**Issue by latenitefilms** [Hammerspoon/hammerspoon#1018]
I'm currently pondering building a some-what crazy feature using Hammerspoon, but before I spend hours trying to make it work, I thought I'd might as well ask the community if you think it's actually possible.

What I want to do, is basically build a "mouse history rewind". What this means is that Hammerspoon will need to be constantly "recording" the mouse movements and presses. If the users presses and holds down a shortcut key, then the mouse will move "back in time", doing the reverse of what it just did until the user stops holding the shortcut key.

Is this remotely possible?

The reason for this is when you're scrolling through lots of different clips in Final Cut Pro, it's really easy to get "lost" as to what you were just looking at. I've considered trying to build a "history list" of the last clips that have been played - but as this would rely on GUI Scripting, I'm not sure how achievable this actually is - however a "mouse history rewind" feature could solve it problem.

Would love to hear what you think! Thanks in advance!

**Comment by latenitefilms** [comment]
@asmagill or @cmsj - Any idea if this is remotely possible or not?

**Comment by cmsj** [comment]
so, it's not impossible, but I tend to imagine that the performance would be fairly dire. A tightly optimised C app might be able to get away with it, but all the jumping back and forth between C and Lua would probably make it fairly painful to live with.

**Comment by asmagill** [comment]
You would need to setup an eventtap and record at least mouse movement and up/down events... perhaps more events as well, you'd have to play with it and see.

There are some issues with eventtaps, though... they introduce a delay in event processing and if Hammerspoon takes too long to handle its recording of the event, macOS will automatically disable the eventtap... we mitigate this by detecting the notification that it's been disabled and re-enable the tap, but... it can introduce apparent slowdowns of the entire system's responsiveness, not just in Hammerspoon.

Their are some tricks that can help, but as @cmsj says, it's hard to do with the jumping back and forth between lua and C... I've tried re-writing the entire hotkey module to use eventtaps because it would allow us finer grained control over modifiers (detect whether or not it was the left or right modifier, use the fn key, etc.) and would allow us to override some of the macOS defaults that we currently can't, but even with almost everything but the callback (i.e. hotkey action) written in C, there were still periods where my keyboard would seem to stop working, then all of a sudden spew the last 10 or 20 characters I'd been typing but not seeing.  I hope to revisit this at some point and see if using multiple threads might help, but the Hammerspoon portion (i.e. Lua callbacks) would still have to be on the main thread, so I don't know if it will ultimately help or not.

Basically, yes, its possible, and I probably would even use Hammerspoon myself to do a proof of concept, but any ultimate solution that I wanted to offer publicly, I'd plan to rewrite in C once I understood the basic requirements.

**Comment by cmsj** [comment]
(I'm going to close this out because I don't think there is any work to be done in Hammerspoon. Please re-open if you disagree!)
